[RF] Merge RooListProxy and RooSetProxy classes into template class

guitargeek opened the pull request:
Merge RooListProxy and RooSetProxy classes into the new template
RooCollectionProxy. This change should cause no backwards compatibility
issues, because the memory layouts remain the same.

hageboeck commented on roofit/roofitcore/inc/RooAbsArg.h:
```diff
@@ -617,8 +619,8 @@ class RooAbsArg : public TNamed, public RooPrintable {
   friend class RooArgProxy ;
-  friend class RooSetProxy ;
-  friend class RooListProxy ;
+  template<class RooCollection_t>
+  friend class RooCollectionProxy;
```
Try without these friend declarations if possible. That's a code smell. Sometimes it was possible to remove them, because old code had already been modernised.

hageboeck commented on roofit/roofitcore/inc/RooSetProxy.h:
```diff
@@ -13,64 +13,266 @@
  *****************************************************************************/
+
+/**
+\class RooSetCollectionProxy
```
```diff
- \class RooSetCollectionProxy
+ \class RooCollectionProxy
```

hageboeck replied:
If you move it directly before the class declaration, the `\class` tag becomes unnecessary. We use it normally in the `.cxx` to keep the headers shorter, but here it's header only.

guitargeek replied:
Okay! I updated this.

hageboeck commented on roofit/roofitcore/inc/RooParamBinning.h:
```diff
@@ -70,7 +70,7 @@ class RooParamBinning : public RooAbsBinning {
   RooAbsReal* xhi() const { return _lp ? ((RooAbsReal*)_lp->at(1)) : _xhi ; }
 
-  ClassDef(RooParamBinning,2) // Binning specification with ranges parameterized by external RooAbsReal functions
+  ClassDef(RooParamBinning,3) // Binning specification with ranges parameterized by external RooAbsReal functions
```
This shouldn't be necessary if the typedefs would be done correctly. Double check them:
the warning message hints at a pointer-vs-object difference.

guitargeek replied:
I triple-checked this, and there is no inconsistency with the `typedefs`. I suppose it's some quirk in IO that the warning appears.

hageboeck reviewed (commented):
Nice try, but you have to be careful with schema evolution.
1. Change the classes, and implement the proper typedefs, so no other code in RF has to be touched.
2. Create schema evolution rules like those:
https://github.com/root-project/root/blob/05d10d2127282c5ec11b1330763dc0b43a93a6df/roofit/roofitcore/inc/LinkDef.h#L185-L192

guitargeek commented:
> Nice try, but you have to be careful with schema evolution.
> 
>     1. Go back, write a class that uses those proxies to a file, commit the file to git, and implement a read test.
> 
>     2. Change the classes, and implement the proper typedefs, so no other code in RF has to be touched.
> 
>     3. Create schema evolution rules like those:
>        https://github.com/root-project/root/blob/05d10d2127282c5ec11b1330763dc0b43a93a6df/roofit/roofitcore/inc/LinkDef.h#L185-L192

Hi @hageboeck, thanks for the review! The proper typedefs were already implemented, and now I also implemented schema evolution rules and implemented a read test.

lmoneta commented on roofit/roofitcore/src/RooArgSet.cxx:
```diff
@@ -104,8 +104,10 @@ void RooArgSet::cleanup()
 {
-  //This will fail if a derived class uses this operator
-  assert(sizeof(RooArgSet) == bytes);
+  // To make sure that derived classes don't use this operator
```
Is not this a different behaviour than before ? 
And before was failing only in debug mode, was this intended ?

guitargeek replied:
Yes, the behavior is different now. Before, the assert would fail if the size doesn't match (i.e. because if the class derives from and extends RooArgSet), now it falls back to not using the memory pool for derived classes.

Before, not using the memory pool for `RooSetProxy` was achieved by overriding `operator new` again in `RooSetProxy`, which is not needed after this change because the `RooArgSet::operator new` will fallback itself to the original operator new for derived classes.

lmoneta reviewed (approved):
Looks good, 
just two small comments. 
Thank you Jonas for this nice improvement !

guitargeek commented on roofit/roofitcore/inc/RooSetProxy.h:
```diff
@@ -13,65 +13,274 @@
  * with or without modification, are permitted according to the terms        *
```
Ok, that's a good point! I have now implemented what you suggested.
